courseworks/week7/ZhigaoXIE/HW7.jl

```diff
@@ -0,0 +1,62 @@
+using Optim
+using ForwardDiff
+using Plots
+using LinearAlgebra
+
+function loss_function(x, edges)
+    n = size(x, 2)
+    loss = 0.0
+    for (u, v) in edges
+        if norm(x[:, u] - x[:, v]) > 1.0
+            loss += norm(x[:, u] - x[:, v]) - 1.0
+        end
+    end
+    for u in 1:n
+        for v in (u + 1):n
+            if (u, v) ∉ edges && norm(x[:, u] - x[:, v]) < 1.0
+                loss += 1.0 - norm(x[:, u] - x[:, v])
+            end
+        end
+    end
+    return loss
+end
+
+function unit_disk_embedding(edges, n)
+    # Assign each vertex with a random coordinate
+    x0 = rand(2, n)
+    res = optimize(x -> loss_function(x, edges), x0, BFGS(), autodiff=:forward)
+
+    # Extract the optimal coordinates from the optimizer result
+    x_opt = res.minimizer
```

Comment:
minimizing the loss function is not equivalent to getting a unit disk distribution, you have to check that, here is an example.
```
function is_unit_disk(x::AbstractArray)
	E = [(1, 2), (1, 3), (2, 3), (2, 4), (2, 5), (2, 6), (3, 5), (3, 6), (3, 7),(4, 5), (4, 8), (5, 6), (5, 8), (5, 9), (6, 7), (6, 8), (6, 9), (7,9), (8, 9), (8, 10), (9, 10)]

	Loss = 0

	for i in 1:10
		for j in i + 1:10
			if (i, j) in E
				d_ij = distance(x, i, j)
				if d_ij >= 1
					Loss += 1
				end
			else
				d_ij = distance(x, i, j)
				if d_ij <= 1
					Loss += 1
				end
			end
		end
	end

	if Loss == 0
		println("The result is a unit disk.")
	else
		println("The result is not a unit disk.")
	end
	return Loss
end
```